core/src/main/kotlin/com/justai/jaicf/api/routing/BotRoutingApi.kt

```diff
@@ -0,0 +1,164 @@
+package com.justai.jaicf.api.routing
+
+import com.justai.jaicf.context.BotContext
+import com.justai.jaicf.context.DefaultActionContext
+import com.justai.jaicf.context.DialogContext
+import com.justai.jaicf.helpers.logging.WithLogger
+import com.justai.jaicf.model.state.StatePath
+import java.util.EmptyStackException
+import java.util.Stack
+
+/**
+ * An exception thrown to re-route current BotRequest to specified [targetEngineName]
+ *
+ * @param [targetEngineName] an engine with specified name provided in [BotRoutingEngine] routables map.
+ * */
+internal data class BotRequestRerouteException(val targetEngineName: String) : RuntimeException()
+
+/**
+ * An exception thrown when there is no previous engine to route back.
+ * */
+internal object NoRouteBackException : RuntimeException()
```

Comment:
1. Same about internal
2. Let's not make a singleton what is not a singleton in semantics